File: exercism/zig/sieve/sieve.zig
const std = @import("std");

pub fn primes(buffer: []u32, limit: u32) []u32 {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    // the first two bits aren't used, the indexes directly correspond to marked numbers
    var sieve = std.DynamicBitSet.initEmpty(allocator, limit + 1) catch |err| @panic(@errorName(err));
    var buff_idx: usize = 0;
    for (2..limit + 1) |n| {
        if (sieve.isSet(n)) continue;
        buffer[buff_idx] = @as(u32, @intCast(n));
        buff_idx += 1;
        // start from n^n as any composite numbers below are divisible by k<n and already marked
        var i = n * n;
        while (i <= limit) : (i += n) {
            sieve.set(i);
        }
    }
    return buffer[0..buff_idx];
}

File: exercism/go/sieve/sieve.go
package sieve

func Sieve(limit int) []int {
	sieve := make([]bool, limit+1) // true - divisible, false - prime
	res := make([]int, 0, limit/2)
	for n := 2; n <= limit; n++ {
		if sieve[n] {
			continue
		}
		res = append(res, n)
		for i := n * 2; i <= limit; i += n {
			sieve[i] = true
		}
	}
	return res
}
